File: native/android/orb/src/main/kotlin/com/earlbalai/orb/OrbOptions.kt
/*
 * OrbOptions.kt
 *
 * The option set, same names and defaults as the web module. Validation is at the
 * boundary: a bad value fails where you wrote it, not a frame later inside a uniform
 * upload.
 */
package com.earlbalai.orb

import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/** How the palette is chosen. */
sealed class OrbPaletteChoice {
  /** Derived from the seed (default). */
  object Auto : OrbPaletteChoice()
  /** Built around a hue in degrees. */
  data class Hue(val degrees: Double) : OrbPaletteChoice()
  /** A hand-built palette. */
  data class Custom(val palette: OrbPalette) : OrbPaletteChoice()
}

/** Archetype option: seed-derived or forced. */
sealed class OrbArchetypeChoice {
  object Auto : OrbArchetypeChoice()
  data class Fixed(val archetype: OrbArchetype) : OrbArchetypeChoice()
}

/** Chromatic rim lens option. */
sealed class OrbLensChoice {
  /** On at or above 48 dp, amplitude tapered by resolution (default). */
  object Auto : OrbLensChoice()
  object Off : OrbLensChoice()
  object On : OrbLensChoice()
  /** Explicit amplitude in p-units, opts out of the taper. Must be >= 0. */
  data class Amplitude(val value: Double) : OrbLensChoice()
}

data class OrbOptions(
  /** Identity. Sets palette, archetype, galaxy structure, meteor cadence and starting spin. */
  val seed: String = "",
  /** Density-independent pixels, square. Minimum 8. */
  val size: Double = 320.0,
  /** Instant at mount, crossfades (~350 ms) afterwards. Continuous on 0..3. */
  val state: Double = 0.0,
  val archetype: OrbArchetypeChoice = OrbArchetypeChoice.Auto,
  val palette: OrbPaletteChoice = OrbPaletteChoice.Auto,
  /**
   * The colour behind the orb, transmitted through the glass. There is no DOM to walk,
   * so set it to your card colour; black by default.
   */
  val background: OrbColor = OrbColor.BLACK,
  /** `false` freezes on the current frame. State changes still repaint. */
  val animate: Boolean = true,
  val lens: OrbLensChoice = OrbLensChoice.Auto,
  /** The inset glass highlight ring. */
  val bevel: Boolean = true,
  /** Render one static frame while the system animator scale is zero. */
  val respectReducedMotion: Boolean = true,
) {
  constructor(seed: String, size: Double = 320.0, state: OrbState = OrbState.IDLE) :
    this(seed = seed, size = size, state = state.index.toDouble())

  /** Throws [IllegalArgumentException] on anything the renderer could not use. */
  fun validate() {
    require(size.isFinite() && size >= 8) { "[Orb] size must be a finite number >= 8" }
    require(state.isFinite() && state in 0.0..3.0) { "[Orb] state must be within 0..3" }
    if (lens is OrbLensChoice.Amplitude) {
      require(lens.value.isFinite() && lens.value >= 0) { "[Orb] lens amplitude must be a finite number >= 0" }
    }
    if (palette is OrbPaletteChoice.Hue) require(palette.degrees.isFinite()) { "[Orb] palette hue must be finite" }
    if (palette is OrbPaletteChoice.Custom) require(palette.palette.accents.size == 3) { "[Orb] palette needs 3 accents" }
  }
}

/** Resolution and lens rules shared by every port. SPEC.md §4. */
internal object OrbResolution {
  const val MAX_PX = 1280
  const val MIN_PX = 8
  const val MAX_DPR = 2.0
  const val HERO_MIN_SIZE = 48.0
  const val LENS_REF_PX = 420.0

  fun pixels(size: Double, density: Double): Int {
    val dpr = if (size >= HERO_MIN_SIZE) min(MAX_DPR, max(density, 1.0)) else 1.0
    return (size * dpr).roundToInt().coerceIn(MIN_PX, MAX_PX)
  }

  fun lens(choice: OrbLensChoice, size: Double, px: Int): Double = when (choice) {
    is OrbLensChoice.Amplitude -> choice.value
    OrbLensChoice.Off -> 0.0
    OrbLensChoice.On -> 0.4 * OrbMath.clamp(LENS_REF_PX / px, 0.55, 1.0)
    OrbLensChoice.Auto -> if (size >= HERO_MIN_SIZE) 0.4 * OrbMath.clamp(LENS_REF_PX / px, 0.55, 1.0) else 0.0
  }
}

File: native/apple/Sources/OrbKit/OrbOptions.swift
//  OrbOptions.swift
//  OrbKit
//
//  The option set, same names and defaults as the web module. Validation is at the
//  boundary: a bad value fails where you wrote it, not a frame later inside a
//  uniform upload.

import Foundation

/// How the palette is chosen.
public enum OrbPaletteChoice: Equatable, Sendable {
  /// Derived from the seed (default).
  case auto
  /// Built around a hue in degrees.
  case hue(Double)
  /// A hand-built palette.
  case custom(OrbPalette)
}

/// Archetype option: seed-derived or forced.
public enum OrbArchetypeChoice: Equatable, Sendable {
  case auto
  case fixed(OrbArchetype)
}

/// Chromatic rim lens option.
public enum OrbLensChoice: Equatable, Sendable {
  /// On at or above 48 pt, amplitude tapered by resolution (default).
  case auto
  case off
  case on
  /// Explicit amplitude in p-units, opts out of the taper. Must be >= 0.
  case amplitude(Double)
}

public struct OrbOptions: Equatable, Sendable {
  /// Identity. Sets palette, archetype, galaxy structure, meteor cadence and starting spin.
  public var seed: String = ""
  /// Points, square. Minimum 8.
  public var size: Double = 320
  /// Instant at mount, crossfades (~350 ms) afterwards. Continuous on 0...3.
  public var state: Double = 0
  public var archetype: OrbArchetypeChoice = .auto
  public var palette: OrbPaletteChoice = .auto
  /// The colour behind the orb, transmitted through the glass. There is no DOM to walk
  /// on a native platform, so set it to your card colour; black by default.
  public var background: OrbColor = .black
  /// `false` freezes on the current frame. State changes still repaint.
  public var animate: Bool = true
  public var lens: OrbLensChoice = .auto
  /// The inset glass highlight ring.
  public var bevel: Bool = true
  /// Render one static frame while the OS reduce-motion setting is on.
  public var respectReducedMotion: Bool = true

  public init() {}

  public init(seed: String, size: Double = 320, state: OrbState = .idle) {
    self.seed = seed
    self.size = size
    self.state = Double(state.rawValue)
  }

  /// Throws on anything the renderer could not use.
  func validate() throws {
    guard size.isFinite, size >= 8 else { throw OrbError.invalid("size must be a finite number >= 8") }
    guard state.isFinite, (0...3).contains(state) else { throw OrbError.invalid("state must be within 0...3") }
    if case .amplitude(let a) = lens, !(a.isFinite && a >= 0) {
      throw OrbError.invalid("lens amplitude must be a finite number >= 0")
    }
    if case .hue(let h) = palette, !h.isFinite { throw OrbError.invalid("palette hue must be finite") }
    if case .custom(let p) = palette, p.accents.count != 3 { throw OrbError.invalid("palette needs 3 accents") }
  }
}

public enum OrbError: Error, CustomStringConvertible {
  case invalid(String)
  case metalUnavailable
  case shaderCompile(String)

  public var description: String {
    switch self {
    case .invalid(let m): return "[Orb] " + m
    case .metalUnavailable: return "[Orb] Metal is unavailable on this device"
    case .shaderCompile(let m): return "[Orb] shader failed to compile: " + m
    }
  }
}

/// Resolution and lens rules shared by every port. SPEC.md §4.
enum OrbResolution {
  static let maxPx = 1280
  static let minPx = 8
  static let maxDpr = 2.0
  static let heroMinSize = 48.0
  static let lensRefPx = 420.0

  static func pixels(size: Double, scale: Double) -> Int {
    let dpr = size >= heroMinSize ? min(maxDpr, max(scale, 1)) : 1
    return Int(OrbMath.clamp((size * dpr).rounded(), Double(minPx), Double(maxPx)))
  }

  static func lens(_ choice: OrbLensChoice, size: Double, px: Int) -> Double {
    switch choice {
    case .amplitude(let a): return a
    case .off: return 0
    case .on: return 0.4 * OrbMath.clamp(lensRefPx / Double(px), 0.55, 1)
    case .auto:
      guard size >= heroMinSize else { return 0 }
      return 0.4 * OrbMath.clamp(lensRefPx / Double(px), 0.55, 1)
    }
  }
}
